jax/_src/dispatch.py

```diff
@@ -544,27 +580,48 @@ def _check_special(name, xla_shape, buf):
     if config.jax_debug_infs and np.any(np.isinf(buf.to_py())):
       raise FloatingPointError(f"invalid value (inf) encountered in {name}")
 
+def _add_tokens(effects: List[core.Effect], device, input_bufs):
+  tokens = [runtime_tokens.get_token(eff, device) for eff in effects]
+  tokens_flat = flatten(tokens)
+  input_bufs = [*tokens_flat, *input_bufs]
+  def _remove_tokens(output_bufs):
+    token_bufs, output_bufs = util.split_list(output_bufs, [len(effects)])
+    for eff, token_buf in zip(effects, token_bufs):
+      runtime_tokens.update_token(eff, token_buf)
+    return output_bufs
+  return input_bufs, _remove_tokens
+
 
 def _execute_compiled(name: str, compiled: XlaExecutable,
                       input_handler: Optional[Callable],
                       output_buffer_counts: Optional[Sequence[int]],
-                      result_handlers, kept_var_idx, *args):
+                      result_handlers,
+                      effects: List[core.Effect],
+                      kept_var_idx, *args):
   device, = compiled.local_devices()
   args = input_handler(args) if input_handler else args
   input_bufs_flat = flatten(device_put(x, device) for i, x in enumerate(args)
                             if i in kept_var_idx)
+  if effects:
```

Comment:
A slight catch is this won't do great things for performance (e.g., building a new closure), but it probably doesn't matter a whole lot since effects aren't yet common. And if we did want them to be fast we should move this logic to C++.

Reply:
Yep I left a TODO to enable this in the fast path.